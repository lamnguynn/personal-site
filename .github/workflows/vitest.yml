name: Run unit and UI tests on project

on:
    push:
      branches:
        - "main"
    pull_request:

jobs:
  run-vitest:
    name: Run vitest
    runs-on: ubuntu-latest
    steps:
      - name:  Checkout code
        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
      - name: Setup Node
        uses: actions/setup-node@49933ea5288caeca8642d1e84afbd3f7d6820020 # v4.4.0
        with:
          node-version: 24
      - name: Install dependencies
        run: |
            npm ci
      - name: Run unit tests
        run: |
          npm run test-unit
      - name: Run UI tests
        run: |
          npm run test-ui