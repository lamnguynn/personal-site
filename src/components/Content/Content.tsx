import type { ContentType } from '@/types/RaceCircuitType';
import { useMemo, type ReactNode } from 'react';

interface Props {
  sectionTitle: string | undefined;
  data: ContentType;
}

export default function Content({ sectionTitle, data }: Props) {
  const content: ReactNode = useMemo(() => {
    const keys = Object.keys(data);
    const index = keys.findIndex((key) => key === sectionTitle);
    const key = Object.keys(data)[0];
    if (index > -1 && sectionTitle) {
      return data[sectionTitle];
    }

    return data[key];
  }, [sectionTitle]);

  return <div className="bg-amber-300 w-full h-full text-sm">{content}</div>;
}
