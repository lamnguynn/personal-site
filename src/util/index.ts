import {
  convertToDegrees,
  convertToRadians,
  getContent,
  getDistanceVector3,
} from './util';

export { convertToDegrees, convertToRadians, getContent, getDistanceVector3 };
