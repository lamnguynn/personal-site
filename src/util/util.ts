import { workData, projectsData, meData } from "@/constants/RaceCircuitConstants";
import * as THREE from "three";
import type { Vector3 } from "three";

export const convertToRadians = (degrees: number) => degrees * (Math.PI / 180);

export const convertToDegrees = (radians: number) => radians * (180 / Math.PI);

/**
 * Get the distance between two given vectors in 3d space
 * 
 * @param val1 
 * @param val2 
 * @returns 
 */
export const getDistanceVector3 = (val1: Vector3 | [number, number, number], val2: Vector3 |  [number, number, number]) => {
    const vectorA: Vector3 = Array.isArray(val1) ? new THREE.Vector3(...val1) : val1;
    const vectorB: Vector3 = Array.isArray(val2) ? new THREE.Vector3(...val2) : val2;
    return vectorA.distanceTo(vectorB);
}

/**
 * Extract both key and values from a certain dataset given an index.
 * 
 * @param index 
 * @returns 
 */
export const getContent = (index: number) => {
    return {
        data: index === 1 ? workData : 
            index === 2 ? projectsData :
            index === 3 ? meData : {},
        keys: index === 1 ?  Object.keys(workData) : 
            index === 2 ?  Object.keys(projectsData) :
            index === 3 ?  Object.keys(meData) : []
    }
}