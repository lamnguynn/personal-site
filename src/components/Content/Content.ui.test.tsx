import { render } from '@testing-library/react';
import { describe, expect, test } from 'vitest';

import { getContent } from '@/util';

import Content from './Content';

describe('Content component', () => {
  test('snapshot', () => {
    const { asFragment } = render(
      <Content sectionTitle={'Section title'} data={getContent(0).data} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
