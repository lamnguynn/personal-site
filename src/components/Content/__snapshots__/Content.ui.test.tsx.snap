// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Content component > snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-amber-300 w-full h-full text-sm"
  />
</DocumentFragment>
`;
